### PR TITLE
TC-166 - adding eks functions

### PR DESCRIPTION
**Pull request by Matan-Amoyal**
Successfully deployed the following service:
https://docs.aws.amazon.com/eks/latest/userguide/eks-guestbook.html

Note:
When using kubectl in new container, it does not have the permissions to communicate with the cluster.
the function "getKubectlConfig" solved it. so anytime we need to execute kubectl command in new containment, need to execute this command in the same container.

### CODE REVIEW

**Comment by Matan-Amoyal** [comment]
All the function tested via Jenkins.

@slavik57 please review.

File: src/main/groovy/aws/eks/EksServices.groovy [hunk 1]
```diff
@@ -0,0 +1,57 @@
...
+        return this.awsCliManager.runAwsCommand(
+            "/bin/bash -c '${this.eksCliManager.getKubectlConfig(this.clusterName)} && kubectl apply -f ${manifestFile}'")
+    }
+
+    def deleteService(serviceName) {
+        return this.awsCliManager.runAwsCommand(
+            "/bin/bash -c '${this.eksCliManager.getKubectlConfig(this.clusterName)} && kubectl delete svc ${serviceName}'")
+    }
+
+    def findServicesWithExternalIP () {
```
Comment (slavik57) [inline src/main/groovy/aws/eks/EksServices.groovy]: How about adding a documentation to the object structure that is returned?

File: src/main/groovy/aws/eks/EksServices.groovy [hunk 2]
```diff
@@ -0,0 +1,57 @@
...
+        return this.awsCliManager.runAwsCommand(
+            "/bin/bash -c '${this.eksCliManager.getKubectlConfig(this.clusterName)} && kubectl get services'")
+    }
+
+    String getServicesInJson() {
+        return this.awsCliManager.runAwsCommand(
+            "/bin/bash -c '${this.eksCliManager.getKubectlConfig(this.clusterName)} && kubectl get services -o json'")
+    }
+
+    def deployService(manifestFile) {
```
Comment (slavik57) [inline src/main/groovy/aws/eks/EksServices.groovy]: Add documentation to the structure of the manifest file.
A link to external documentation is also acceptable

File: src/main/groovy/aws/eks/EksCliManager.groovy [hunk 3]
```diff
@@ -17,4 +17,8 @@ class EksCliManager implements Serializable {
     def getKubectlVersion() {
         return this.awsCliManager.runAwsCommand("kubectl version --client")
     }
+
+    String getKubectlConfig(clusterName) {
```
Comment (slavik57) [inline src/main/groovy/aws/eks/EksCliManager.groovy]: rename to `getUpdateKubeconfigCommand`

File: src/main/groovy/aws/eks/EksCluster.groovy [hunk 4]
```diff
@@ -0,0 +1,38 @@
...
+    def awsCliManager
+    def eksCliManager
+
+    EksCluster(pipeline, awsCliManager, eksCliManager) {
+        this.pipeline = pipeline
+        this.awsCliManager = awsCliManager
+        this.eksCliManager = eksCliManager
+    }
+
+    def createCluster(clusterConfigurationFilePath){
```
Comment (slavik57) [inline src/main/groovy/aws/eks/EksCluster.groovy]: Add documentation to the structure of the file.
A link to external documentation is also acceptable

File: src/main/groovy/aws/eks/EksCluster.groovy [hunk 5]
```diff
@@ -0,0 +1,38 @@
...
+        return this.awsCliManager.runAwsCommand("eksctl create cluster --profile ${this.awsCliManager.awsProfie} -f ${clusterConfigurationFilePath}", [clusterConfigurationFilePath])
+    }
+
+    def getClusters() {
```
Comment (slavik57) [inline src/main/groovy/aws/eks/EksCluster.groovy]: Add documentation to the structure of the retuned object

File: src/main/groovy/aws/eks/EksCluster.groovy [hunk 6]
```diff
@@ -0,0 +1,38 @@
...
+        def eksServices = new EksServices(this.pipeline, this.awsCliManager, this.eksCliManager, clusterName)
+        def servicesWithExternalIP = eksServices.findServicesWithExternalIP()
+        for (service in servicesWithExternalIP) {
+                eksServices.deleteService(service)
+        }
+        
+        return this.awsCliManager.runAwsCommand(
+            "eksctl delete cluster --name ${clusterName} --profile ${this.awsCliManager.awsProfie}")
+    }
+
```
Comment (slavik57) [inline src/main/groovy/aws/eks/EksCluster.groovy]: redundant enter

File: src/main/groovy/aws/eks/EksServices.groovy [hunk 7]
```diff
@@ -0,0 +1,57 @@
...
+        this.clusterName = clusterName
+        this.jsonUtils = new JSONUtils(this.pipeline)
+    }
+
+    def getReplicationcontrollers() {
+        return this.awsCliManager.runAwsCommand(
+            "/bin/bash -c '${this.eksCliManager.getKubectlConfig(this.clusterName)} && kubectl get rc'")
+    }
+    
+    def getServices() {
```
Comment (slavik57) [inline src/main/groovy/aws/eks/EksServices.groovy]: Add documentation to the structure of the returned object.
A link to external documentation is also possible

File: src/main/groovy/aws/eks/EksServices.groovy [hunk 8]
```diff
@@ -0,0 +1,57 @@
...
+        return this.awsCliManager.runAwsCommand(
+            "/bin/bash -c '${this.eksCliManager.getKubectlConfig(this.clusterName)} && kubectl get services'")
+    }
+
+    String getServicesInJson() {
```
Comment (slavik57) [inline src/main/groovy/aws/eks/EksServices.groovy]: Add documentation to the structure of the returned object.
A link to external documentation is also possible

File: src/main/groovy/aws/eks/EksServices.groovy [hunk 9]
```diff
@@ -0,0 +1,57 @@
...
+
+    EksServices(pipeline, awsCliManager, eksCliManager, clusterName) {
+        this.pipeline = pipeline
+        this.awsCliManager = awsCliManager
+        this.eksCliManager = eksCliManager
+        this.clusterName = clusterName
+        this.jsonUtils = new JSONUtils(this.pipeline)
+    }
+
+    def getReplicationcontrollers() {
```
Comment (slavik57) [inline src/main/groovy/aws/eks/EksServices.groovy]: Add documentation to the structure of the returned object.
A link to external documentation is also possible

**Comment by Matan-Amoyal** [comment]
> Did you make sure that all the created resources are deleted?
> (Security groups, Private networks, DNS names etc)

Yes, the cluster created via CloudFormation. It create a stack and when I request to delete the cluster, it delete the stack from CloudFormation.

The only issue is when delete a cluster with a service with external IP. so the solution is to delete first the service and then delete the cluster. (as document in AWS documents). - tested by me and it worked

**Comment by Matan-Amoyal** [comment]
Fixed. @slavik57 review

File: src/main/groovy/aws/eks/EksCluster.groovy [hunk 10]
```diff
@@ -0,0 +1,39 @@
...
+        this.awsCliManager = awsCliManager
+        this.eksCliManager = eksCliManager
+    }
+
+    // Example of config file: https://eksctl.io/
+    def createCluster(clusterConfigurationFilePath){
+        return this.awsCliManager.runAwsCommand("eksctl create cluster --profile ${this.awsCliManager.awsProfie} -f ${clusterConfigurationFilePath}", [clusterConfigurationFilePath])
+    }
+
+    // The function return a table (Name, Region) of all active clusters
```
Comment (slavik57) [inline src/main/groovy/aws/eks/EksCluster.groovy]: This is not really clear, how about giving an example of a result

File: src/main/groovy/aws/eks/EksServices.groovy [hunk 11]
```diff
@@ -0,0 +1,64 @@
...
+
+    EksServices(pipeline, awsCliManager, eksCliManager, clusterName) {
+        this.pipeline = pipeline
+        this.awsCliManager = awsCliManager
+        this.eksCliManager = eksCliManager
+        this.clusterName = clusterName
+        this.jsonUtils = new JSONUtils(this.pipeline)
+    }
+
+    // The function return a table (Name, Desired, current, ready) of replicationcontrollers in the cluster
```
Comment (slavik57) [inline src/main/groovy/aws/eks/EksServices.groovy]: Give an example

File: src/main/groovy/aws/eks/EksServices.groovy [hunk 12]
```diff
@@ -0,0 +1,64 @@
...
+    def getReplicationcontrollers() {
+        return this.awsCliManager.runAwsCommand(
+            "/bin/bash -c '${this.eksCliManager.getUpdateKubeconfigCommand(this.clusterName)} && kubectl get rc'")
+    }
+    
+    // The function return a table (Name, Tyep, cluster-ip, external-ip, ports, age) of services in the cluster
```
Comment (slavik57) [inline src/main/groovy/aws/eks/EksServices.groovy]: Give an example

File: src/main/groovy/aws/eks/EksServices.groovy [hunk 13]
```diff
@@ -0,0 +1,64 @@
...
+    def getServices() {
+        return this.awsCliManager.runAwsCommand(
+            "/bin/bash -c '${this.eksCliManager.getUpdateKubeconfigCommand(this.clusterName)} && kubectl get services'")
+    }
+
+    // The function return a json of services in the cluster
```
Comment (slavik57) [inline src/main/groovy/aws/eks/EksServices.groovy]: Not clear, give an example

File: src/main/groovy/aws/eks/EksServices.groovy [hunk 14]
```diff
@@ -0,0 +1,64 @@
...
+    String getServicesInJson() {
+        return this.awsCliManager.runAwsCommand(
+            "/bin/bash -c '${this.eksCliManager.getUpdateKubeconfigCommand(this.clusterName)} && kubectl get services -o json'")
+    }
+
+    // This function get a manifest file (can be for service, rc and etc..) as a web link or path to file
+    // for example the following links into this document
+    // https://docs.aws.amazon.com/eks/latest/userguide/eks-guestbook.html
```
Comment (slavik57) [inline src/main/groovy/aws/eks/EksServices.groovy]: The link leads to a manual contains a lot of things, please pinpoint to the exact place where to look there

File: src/main/groovy/aws/eks/EksCluster.groovy [hunk 15]
```diff
@@ -0,0 +1,39 @@
...
+    def getClusters() {
+        return this.awsCliManager.runAwsCommand(
+            "eksctl get cluster --profile ${this.awsCliManager.awsProfie}")
+    }
+    
+    def deleteCluster(clusterName) {
+        // First need to delete all services with external-ip value, for more details:
+        // https://docs.aws.amazon.com/eks/latest/userguide/delete-cluster.html
```
Comment (slavik57) [inline src/main/groovy/aws/eks/EksCluster.groovy]: This page contains a lot of things, please pinpoint where to look there as well

File: src/main/groovy/aws/eks/EksServices.groovy [hunk 16]
```diff
@@ -0,0 +1,64 @@
...
+        return this.awsCliManager.runAwsCommand(
+            "/bin/bash -c '${this.eksCliManager.getUpdateKubeconfigCommand(this.clusterName)} && kubectl apply -f ${manifestFile}'")
+    }
+
+    def deleteService(serviceName) {
+        return this.awsCliManager.runAwsCommand(
+            "/bin/bash -c '${this.eksCliManager.getUpdateKubeconfigCommand(this.clusterName)} && kubectl delete svc ${serviceName}'")
+    }
+
+    // This function searching for services with external ip. and return array of the services names.
```
Comment (slavik57) [inline src/main/groovy/aws/eks/EksServices.groovy]: Give an example

File: src/main/groovy/aws/eks/EksCluster.groovy (hunk 15 above)
Comment (Matan-Amoyal) [reply src/main/groovy/aws/eks/EksCluster.groovy]: The page contain the exactly what we need in this function. (get services, delete service with external IP and delete cluster). I can't pinpoint it more then this page

**Comment by Matan-Amoyal** [comment]
I amend the new commit to previous one.
I handled every comment except the one about 'delete cluster' document.

@slavik57

File: src/main/groovy/aws/eks/EksServices.groovy [hunk 17]
```diff
@@ -0,0 +1,107 @@
...
+        return this.awsCliManager.runAwsCommand(
+            "/bin/bash -c '${this.eksCliManager.getUpdateKubeconfigCommand(this.clusterName)} && kubectl apply -f ${manifestFile}'")
+    }
+
+    def deleteService(serviceName) {
+        return this.awsCliManager.runAwsCommand(
+            "/bin/bash -c '${this.eksCliManager.getUpdateKubeconfigCommand(this.clusterName)} && kubectl delete svc ${serviceName}'")
+    }
+
+    // This function searching for services with external ip. and return array of the services names:
```
Comment (slavik57) [inline src/main/groovy/aws/eks/EksServices.groovy]: This is not the output of this function, it returns a list of strings from what I understand

**Comment by Matan-Amoyal** [comment]
My bad, too much monitors for one day :)

**Comment by Matan-Amoyal** [comment]
@slavik57 please review again

File: src/main/groovy/aws/eks/EksServices.groovy (hunk 14 above)
Comment (slavik57) [reply src/main/groovy/aws/eks/EksServices.groovy]: @Matan-Amoyal 
What about this one?

Comment (slavik57) [reply src/main/groovy/aws/eks/EksServices.groovy]: Ignore this comment

**Pull request merged**